File: snippets/Basics/kotlin/factorial.kt
// Metadata Header (MANDATORY)
// -----------------------------
// Program Title: Factorial Calculation (Iterative)
// Author: IamBisrutPyne
// Date: 2025-10-10
//
// Description: Computes the factorial of a non-negative integer using an iterative approach.
//
// Language: Kotlin
//
// Time Complexity: O(n)
// Space Complexity: O(1)
// -----------------------------

/**
 * Calculates the factorial of a non-negative integer.
 * @param n The number to calculate the factorial of.
 * @return The factorial result.
 */
fun factorial(n: Int): Long {
    // Use Long to handle larger results safely
    if (n < 0) {
        return 0
    }
    var result: Long = 1
    for (i in 2..n) {
        result *= i
    }
    return result
}

fun main() {
    val num = 10
    println("Factorial of $num is: ${factorial(num)}")
    // Output: Factorial of 10 is: 3628800
}

File: snippets/Basics/rust/factorial.rs
// Metadata Header (MANDATORY)
// -----------------------------
// Program Title: Factorial Calculation (Iterative)
// Author: IamBisrutPyne
// Date: 2025-10-10
//
// Description: Computes the factorial of a non-negative integer using an iterative approach.
//
// Language: Rust
//
// Time Complexity: O(n)
// Space Complexity: O(1)
// -----------------------------

fn factorial(n: u64) -> u64 {
    if n == 0 {
        return 1;
    }
    let mut result: u64 = 1;
    for i in 2..=n {
        result *= i;
    }
    result
}

fn main() {
    let num: u64 = 19;
    println!("Factorial of {} is: {}", num, factorial(num));
    // Output: Factorial of 10 is: 3628800
}
